### PR TITLE
swss-common: add async methods

### PR DESCRIPTION
**Pull request by erer1243**
Counterpart pr to https://github.com/sonic-net/sonic-swss-common/pull/945
introduces those methods and an async option to await readability on swss channels.
The read_data method would be the most egregious blocking call when using swss-common with tokio, but there may be others. 

I don't know enough about performance characteristics of redis or zmq to say if, for example, writing a field with a ProducerStateTable or connecting to a socket with ZmqClient/Server will block enough to mess with tokio. Rule of thumb for tokio is that there should be no more than 100 microseconds between await points. If they become a problem, we can use `tokio::task::spawn_blocking`. By default, tokio will allow 512 blocking `spawn_blocking` tasks on rt-multi (this can be configured with `tokio::runtime::Builder`, but 512 seems like enough probably. Will we ever have >512 concurrent database writes, or something like that?).

### CODE REVIEW

File: crates/swss-common/tests/sync.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
+mod common;
+
+use common::*;
+use std::{collections::HashMap, time::Duration};
+use swss_common::*;
+
+#[test]
+fn dbconnector() {
```
Comment (r12f) [inline crates/swss-common/tests/sync.rs:8]: dbconnector_sync_api_can_work

Comment (r12f) [reply crates/swss-common/tests/sync.rs:8]: same as other test functions.

File: crates/swss-common/tests/async.rs [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+// This macro verifies that async test functions are Send. tokio::test is a bit misleading because
+// a tokio runtime's root future can be !Send. This takes a test function and defines two tests from
+// it - one that is the root future (Send not required), and one that uses a type assertion to check
+// that the tested future is Send.
+macro_rules! define_tokio_test_fns {
+    ($f:ident) => {
+        paste! {
+            #[tokio::test]
+            async fn [< $f _async >]() {
```
Comment (r12f) [inline crates/swss-common/tests/async.rs:23]: did rust add support for this ident join finally?

Comment (erer1243) [reply crates/swss-common/tests/async.rs:23]: no it's a library :^)

File: crates/swss-common/tests/async.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            fn [< $f _is_send >]() {
+                fn assert_is_send<T: Send>(_: T) {}
+                assert_is_send($f());
+            }
+        }
+    };
+}
+
+define_tokio_test_fns!(dbconnector);
+async fn dbconnector() {
```
Comment (r12f) [inline crates/swss-common/tests/async.rs:37]: same as sync function.

dbconnector_async_api_can_work

File: crates/swss-common/tests/sync.rs (hunk 1 above)
Comment (erer1243) [reply crates/swss-common/tests/sync.rs:8]: "sync" is already in the name because it prints `sync::dbconnector` when you run the tests

Comment (erer1243) [reply crates/swss-common/tests/sync.rs:8]: wait, sorry, no it doesn't anymore. it used to.

**Review by r12f — APPROVED** [review]

**Pull request merged**